backends/fs/fs.go

```diff
@@ -89,6 +90,25 @@ func (b *Backend) Delete(ctx context.Context, name string) error {
 	return err
 }
 
+// NewReader provides an optimized way to read from named file.
+func (b *Backend) NewReader(ctx context.Context, name string) (io.ReadCloser, error) {
+	if !allowedName(name) {
+		return nil, os.ErrPermission
+	}
+	fullPath := filepath.Join(b.rootPath, name)
+	return os.Open(fullPath)
+}
+
+
+// NewWriter provides an optimized way to write to a file.
+func (b *Backend) NewWriter(ctx context.Context, name string) (io.WriteCloser, error) {
+	if !allowedName(name) {
+		return nil, os.ErrPermission
+	}
+	fullPath := filepath.Join(b.rootPath, name)
+	return os.OpenFile(fullPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
```

Comment:
Directly returning the file here makes it immediately visible to List, even if not fully written. This should have the same atomic guarantees as writing to S3: the file only exists if it has been completely written. This means that we need to write the data to a temp file in the same directory that will not be visible in List, and then rename to the desired filename on Close.

Basically wrap it, unlink on error and rename in Close. Do not use `os.CreateTemp` for this, because that has undesired side effects like using the wrong permissions. Instead, construct a temp file name yourself.

The tests should check for this (create a writer, then verify that it does not exist yet in List).